### PR TITLE
Robustness to driver pod taking time to create

### PR DESCRIPTION
**Pull request by Tom-Newton**
## Purpose of this PR
Improve reliability. 
Closes: #2302

**Proposed changes:**
- Add a grace period for the driver pod to be created. 
- Grace period is controlled by a new config option `driver-pod-creation-grace-period`. Default is 10 seconds. 
- Add 3 new tests around reconciling submitted status, including the new grace period functionality. This is the majority of the lines changed. 
- Expose the new config option in the helm chart and add a helm test

## Change Category
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] Feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that could affect existing functionality)
- [ ] Documentation update

## Checklist

- [x] I have conducted a self-review of my own code.
- [x] I have updated documentation accordingly - I think its sufficient just to add a description on the new argument on the helm chart and regenerate the helm docs. 
- [x] I have added tests that prove my changes are effective or that my feature works.
- [x] Existing unit tests pass locally with my changes.

### Additional Notes
Some logs of a real example on our prod cluster where the spark application was saved by the grace period added in this PR. 
[Explore-logs-2024-11-11 12_09_13.txt](https://github.com/user-attachments/files/17702165/Explore-logs-2024-11-11.12_09_13.txt)

### CODE REVIEW

File: internal/controller/sparkapplication/controller.go [hunk 1]
```diff
@@ -773,10 +775,13 @@ func (r *Reconciler) updateDriverState(_ context.Context, app *v1beta2.SparkAppl
 	}
 
 	if driverPod == nil {
-		app.Status.AppState.State = v1beta2.ApplicationStateFailing
-		app.Status.AppState.ErrorMessage = "driver pod not found"
-		app.Status.TerminationTime = metav1.Now()
-		return nil
+		if metav1.Now().Sub(app.Status.LastSubmissionAttemptTime.Time) > r.options.DriverPodCreationGracePeriod {
```
Comment (ChenYi015) [inline internal/controller/sparkapplication/controller.go:778]:
```suggestion
		if app.Status.AppState.State != v1beta2.ApplicationStateSubmitted || metav1.Now().Sub(app.Status.LastSubmissionAttemptTime.Time) > r.options.DriverPodCreationGracePeriod {
```

I think we need to consider the driver pod creation grace period only when the application state is submitted.

File: charts/spark-operator-chart/tests/controller/deployment_test.yaml [hunk 2]
```diff
@@ -637,6 +637,14 @@ tests:
...
           content: --workqueue-ratelimiter-max-delay=1h
+  - it: Should contain `driver-pod-creation-grace-period` arg if `controller.driverPodCreationGracePeriod` is set
+    set:
+      controller:
+        driverPodCreationGracePeriod: 30s
+    asserts:
+      - contains:
+          path: spec.template.spec.containers[?(@.name=="spark-operator-controller")].args
+          content: --driver-pod-creation-grace-period=30s
   - it: Should contain `--max-tracked-executor-per-app` arg if `controller.maxTrackedExecutorPerApp` is set
```
[the remark this answers is not in the document]
Comment (Tom-Newton) [reply charts/spark-operator-chart/tests/controller/deployment_test.yaml:650]: Fixed https://github.com/kubeflow/spark-operator/pull/2315/commits/37846be96a4de6ec2458baad3535801da8ce31be

**Comment by Tom-Newton** [comment]
The CI failure looks unrelated 
```
• [FAILED] [300.030 seconds]
Example SparkApplication spark-pi-python [It] Should complete successfully
/home/runner/work/spark-operator/spark-operator/test/e2e/sparkapplication_test.go:383

  [FAILED] Unexpected error:
      <*fmt.wrapError | 0xc001b63360>: 
      client rate limiter Wait returned an error: context deadline exceeded
      {
          msg: "client rate limiter Wait returned an error: context deadline exceeded",
          err: <context.deadlineExceededError>{},
      }
  occurred
  In [It] at: /home/runner/work/spark-operator/spark-operator/test/e2e/sparkapplication_test.go:386 @ 12/04/24 12:25:47.015
```
I'll try rebasing on latest main

**Review by ChenYi015 — APPROVED** [review]

**Comment by ChenYi015** [comment]
@Tom-Newton Thanks for improving the robustness of Spark operator.

**Comment by ChenYi015** [comment]
@Tom-Newton Thanks for improving the robustness of Spark operator.
/lgtm

**Comment by Tom-Newton** [comment]
> @Tom-Newton Thanks for improving the robustness of Spark operator.

Thanks for reviewing :slightly_smiling_face:

**Pull request merged**